File: typescript/src/streaming/VideoCallService.ts
/**
 * Point-to-point video call service.
 *
 * Mirrors VoiceCallService's state machine but uses:
 *   PacketType.VideoSignaling  (JSON VideoSignalingMessage)
 *   PacketType.VideoFrame      (binary VideoFrame)
 *
 * Extra signaling operations:
 *   requestKeyframe(callId)
 *   notifyQualityChange(callId, width, height, fps, bitrateKbps)
 *
 * Binary VideoFrame:
 *   [16] CallId (UUID RFC4122 big-endian)
 *   [4]  Sequence (uint32 little-endian)
 *   [8]  TimestampMs (int64 little-endian)
 *   [1]  IsKeyframe (0 or 1)
 *   [N]  EncodedPayload
 *
 * SPDX-License-Identifier: MIT
 */

import { DEFAULT_TTL } from "../constants.js";
import { MeshPacket } from "../protocol/MeshPacket.js";
import { PacketType } from "../protocol/PacketType.js";
import { IMeshSender } from "../routing/IMeshSender.js";
import { uuidToBytes, bytesToUuid } from "../voice/VoiceCallService.js";

// ──────────────────────────────────────────────────────────────────
// Constants
// ──────────────────────────────────────────────────────────────────

const VIDEO_CALL_TIMEOUT_SECONDS = 30;
const VIDEO_FRAME_PRIORITY = 64;
const VIDEO_SIGNALING_PRIORITY = 32;

// ──────────────────────────────────────────────────────────────────
// Wire types (snake_case)
// ──────────────────────────────────────────────────────────────────

type VideoSignalingKind =
  | "offer"
  | "answer"
  | "hangup"
  | "keyframe_request"
  | "quality_change";

interface VideoSignalingMessage {
  kind: VideoSignalingKind;
  call_id: string;
  from_uhid: string;
  to_uhid: string;
  proposed_codecs?: string[];
  selected_codec?: string;
  width?: number;
  height?: number;
  fps?: number;
  bitrate_kbps?: number;
  reason?: string;
}

// ──────────────────────────────────────────────────────────────────
// Public types
// ──────────────────────────────────────────────────────────────────

export type VideoCallState = "idle" | "ringing" | "connected" | "ended" | "failed";

export interface VideoCallInfo {
  callId: string;
  remoteUhid: string;
  state: VideoCallState;
  isOutgoing: boolean;
  startedAt: Date;
}

export interface VideoFrameEvent {
  callId: string;
  fromUhid: string;
  sequence: number;
  timestampMs: number;
  isKeyframe: boolean;
  encodedPayload: Uint8Array;
}

export interface VideoQualityParams {
  width: number;
  height: number;
  fps: number;
  bitrateKbps: number;
}

// ──────────────────────────────────────────────────────────────────
// Service
// ──────────────────────────────────────────────────────────────────

export class VideoCallService {
  private readonly calls = new Map<string, VideoCallInfo>();
  private readonly timeouts = new Map<string, ReturnType<typeof setTimeout>>();

  private videoFrameSeq = 0;

  // Event callbacks
  onIncomingCall?: (info: VideoCallInfo) => void;
  onCallConnected?: (info: VideoCallInfo) => void;
  onCallEnded?: (info: VideoCallInfo) => void;
  onFrameReceived?: (event: VideoFrameEvent) => void;
  onKeyframeRequested?: (callId: string, fromUhid: string) => void;
  onQualityChanged?: (callId: string, fromUhid: string, params: VideoQualityParams) => void;

  constructor(private readonly sender: IMeshSender) {}

  // ──────────────── outbound signaling ────────────────────────────

  async sendOffer(toUhid: string, codecs: string[]): Promise<string> {
    if (!toUhid) throw new Error("toUhid must not be empty");

    const callId = crypto.randomUUID();
    const info: VideoCallInfo = {
      callId,
      remoteUhid: toUhid,
      state: "ringing",
      isOutgoing: true,
      startedAt: new Date(),
    };
    this.calls.set(callId, info);
    this.armTimeout(callId);

    await this.sendSignaling({
      kind: "offer",
      call_id: callId,
      from_uhid: this.sender.localUhid,
      to_uhid: toUhid,
      proposed_codecs: codecs,
    }, toUhid);

    return callId;
  }

  async acceptCall(callId: string, selectedCodec?: string): Promise<void> {
    const info = this.calls.get(callId);
    if (!info) return;
    this.clearTimeout(callId);
    info.state = "connected";
    this.onCallConnected?.(info);

    await this.sendSignaling({
      kind: "answer",
      call_id: callId,
      from_uhid: this.sender.localUhid,
      to_uhid: info.remoteUhid,
      selected_codec: selectedCodec,
    }, info.remoteUhid);
  }

  async hangUp(callId: string): Promise<void> {
    const info = this.calls.get(callId);
    if (!info) return;
    this.clearTimeout(callId);
    info.state = "ended";
    this.onCallEnded?.(info);

    await this.sendSignaling({
      kind: "hangup",
      call_id: callId,
      from_uhid: this.sender.localUhid,
      to_uhid: info.remoteUhid,
    }, info.remoteUhid);
  }

  async requestKeyframe(callId: string): Promise<void> {
    const info = this.calls.get(callId);
    if (!info || info.state !== "connected") return;

    await this.sendSignaling({
      kind: "keyframe_request",
      call_id: callId,
      from_uhid: this.sender.localUhid,
      to_uhid: info.remoteUhid,
    }, info.remoteUhid);
  }

  async notifyQualityChange(
    callId: string,
    width: number,
    height: number,
    fps: number,
    bitrateKbps: number,
  ): Promise<void> {
    const info = this.calls.get(callId);
    if (!info || info.state !== "connected") return;

    await this.sendSignaling({
      kind: "quality_change",
      call_id: callId,
      from_uhid: this.sender.localUhid,
      to_uhid: info.remoteUhid,
      width,
      height,
      fps,
      bitrate_kbps: bitrateKbps,
    }, info.remoteUhid);
  }

  // ──────────────── video frame sending ───────────────────────────

  async sendFrame(
    callId: string,
    encodedVideo: Uint8Array,
    isKeyframe: boolean,
  ): Promise<boolean> {
    const info = this.calls.get(callId);
    if (!info || info.state !== "connected") return false;

    const payload = encodeVideoFrame(callId, encodedVideo, isKeyframe, this.videoFrameSeq++);

    const packet = new MeshPacket();
    packet.type = PacketType.VideoFrame;
    packet.sourceUhid = this.sender.localUhid;
    packet.destinationUhid = info.remoteUhid;
    packet.ttl = DEFAULT_TTL;
    packet.priority = VIDEO_FRAME_PRIORITY;
    packet.payload = payload;

    return this.sender.send(packet, info.remoteUhid);
  }

  // ──────────────── inbound packet handling ───────────────────────

  async onPacket(packet: MeshPacket): Promise<void> {
    switch (packet.type) {
      case PacketType.VideoSignaling:
        await this.handleSignaling(packet);
        break;
      case PacketType.VideoFrame:
        this.handleFrame(packet);
        break;
      default:
        break;
    }
  }

  // ──────────────── inspection ─────────────────────────────────────

  getCallInfo(callId: string): VideoCallInfo | undefined {
    return this.calls.get(callId);
  }

  getActiveCalls(): VideoCallInfo[] {
    return Array.from(this.calls.values()).filter(
      (c) => c.state === "ringing" || c.state === "connected",
    );
  }

  // ──────────────── private helpers ─────────────────────────────────

  private async handleSignaling(packet: MeshPacket): Promise<void> {
    let msg: VideoSignalingMessage;
    try {
      msg = JSON.parse(
        new TextDecoder().decode(packet.payload),
      ) as VideoSignalingMessage;
    } catch {
      return;
    }
    if (!msg.call_id || !msg.kind) return;

    const { kind, call_id: callId, from_uhid: fromUhid } = msg;

    switch (kind) {
      case "offer": {
        if (this.calls.has(callId)) return;
        const info: VideoCallInfo = {
          callId,
          remoteUhid: fromUhid,
          state: "ringing",
          isOutgoing: false,
          startedAt: new Date(),
        };
        this.calls.set(callId, info);
        this.armTimeout(callId);
        this.onIncomingCall?.(info);
        break;
      }
      case "answer": {
        const info = this.calls.get(callId);
        if (!info || info.state !== "ringing") break;
        this.clearTimeout(callId);
        info.state = "connected";
        this.onCallConnected?.(info);
        break;
      }
      case "hangup": {
        const info = this.calls.get(callId);
        if (!info) break;
        this.clearTimeout(callId);
        info.state = "ended";
        this.onCallEnded?.(info);
        break;
      }
      case "keyframe_request": {
        const info = this.calls.get(callId);
        if (!info || info.state !== "connected") break;
        this.onKeyframeRequested?.(callId, fromUhid);
        break;
      }
      case "quality_change": {
        const info = this.calls.get(callId);
        if (!info || info.state !== "connected") break;
        this.onQualityChanged?.(callId, fromUhid, {
          width: msg.width ?? 0,
          height: msg.height ?? 0,
          fps: msg.fps ?? 0,
          bitrateKbps: msg.bitrate_kbps ?? 0,
        });
        break;
      }
      default:
        break;
    }
  }

  private handleFrame(packet: MeshPacket): void {
    const frame = decodeVideoFrame(packet.payload);
    if (!frame) return;
    const info = this.calls.get(frame.callId);
    if (!info || info.state !== "connected") return;

    this.onFrameReceived?.({
      callId: frame.callId,
      fromUhid: packet.sourceUhid,
      sequence: frame.sequence,
      timestampMs: frame.timestampMs,
      isKeyframe: frame.isKeyframe,
      encodedPayload: frame.encodedPayload,
    });
  }

  private async sendSignaling(msg: VideoSignalingMessage, toUhid: string): Promise<void> {
    const body = new TextEncoder().encode(JSON.stringify(msg));
    const packet = new MeshPacket();
    packet.type = PacketType.VideoSignaling;
    packet.sourceUhid = this.sender.localUhid;
    packet.destinationUhid = toUhid;
    packet.ttl = DEFAULT_TTL;
    packet.priority = VIDEO_SIGNALING_PRIORITY;
    packet.payload = body;
    await this.sender.send(packet, toUhid);
  }

  private armTimeout(callId: string): void {
    this.clearTimeout(callId);
    const handle = setTimeout(async () => {
      const info = this.calls.get(callId);
      if (!info || info.state !== "ringing") return;
      info.state = "failed";
      this.onCallEnded?.(info);
      await this.sendSignaling({
        kind: "hangup",
        call_id: callId,
        from_uhid: this.sender.localUhid,
        to_uhid: info.remoteUhid,
        reason: "timeout",
      }, info.remoteUhid);
    }, VIDEO_CALL_TIMEOUT_SECONDS * 1000);
    this.timeouts.set(callId, handle);
  }

  private clearTimeout(callId: string): void {
    const handle = this.timeouts.get(callId);
    if (handle !== undefined) {
      clearTimeout(handle);
      this.timeouts.delete(callId);
    }
  }
}

// ──────────────────────────────────────────────────────────────────
// Binary codec
// ──────────────────────────────────────────────────────────────────

/**
 * VideoFrame binary payload:
 * [16] CallId (UUID RFC4122 big-endian)
 * [4]  Sequence (uint32 little-endian)
 * [8]  TimestampMs (int64 little-endian)
 * [1]  IsKeyframe (0 or 1)
 * [N]  EncodedPayload
 */
export function encodeVideoFrame(
  callId: string,
  encodedPayload: Uint8Array,
  isKeyframe: boolean,
  sequence: number,
): Uint8Array {
  const buf = new Uint8Array(16 + 4 + 8 + 1 + encodedPayload.length);
  const dv = new DataView(buf.buffer);

  uuidToBytes(callId, buf, 0);
  dv.setUint32(16, sequence >>> 0, true);
  dv.setBigInt64(20, BigInt(Date.now()), true);
  buf[28] = isKeyframe ? 1 : 0;
  buf.set(encodedPayload, 29);

  return buf;
}

export function decodeVideoFrame(data: Uint8Array): {
  callId: string;
  sequence: number;
  timestampMs: number;
  isKeyframe: boolean;
  encodedPayload: Uint8Array;
} | null {
  if (data.length < 29) return null;
  const dv = new DataView(data.buffer, data.byteOffset, data.byteLength);

  const callId = bytesToUuid(data, 0);
  const sequence = dv.getUint32(16, true);
  const timestampMs = Number(dv.getBigInt64(20, true));
  const isKeyframe = data[28] !== 0;
  const encodedPayload = data.slice(29);

  return { callId, sequence, timestampMs, isKeyframe, encodedPayload };
}

File: kotlin/src/main/kotlin/aethernet/streaming/VideoCallService.kt
// SPDX-License-Identifier: MIT

package aethernet.streaming

import aethernet.AetherNetConstants
import aethernet.protocol.MeshPacket
import aethernet.protocol.PacketType
import aethernet.routing.MeshSender
import aethernet.voice.JsonReader
import aethernet.voice.jsonEscape
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

// ─────────────────────────────────────────────────────────────────────────────
// Domain types
// ─────────────────────────────────────────────────────────────────────────────

enum class VideoCallState { Outgoing, Incoming, Connected, Ended, Failed }

data class VideoCallSession(
    val id: UUID,
    val callerUhid: String,
    val calleeUhid: String,
    var state: VideoCallState,
    val proposedCodecs: List<String> = emptyList(),
    var selectedCodec: String? = null,
    var width: Int = 0,
    var height: Int = 0,
    var fps: Int = 0,
    var bitrateKbps: Int = 0
)

/**
 * One-to-one video call service.
 *
 * Wire format — VideoFrame binary payload:
 *   [16] CallId      (UUID RFC4122 big-endian)
 *   [4]  Sequence    (UInt32 little-endian)
 *   [8]  TimestampMs (Int64 little-endian)
 *   [1]  IsKeyframe  (0 or 1)
 *   [N]  EncodedPayload
 *
 * Signaling uses VideoSignalingMessage JSON (snake_case).
 *
 * Priority: 64 for video frames, 32 for signaling.
 */
class VideoCallService(private val sender: MeshSender) {

    private val sessions = ConcurrentHashMap<UUID, VideoCallSession>()
    @Volatile private var frameSequence: Int = 0

    /** Fired when an incoming video call offer arrives. */
    var onIncomingCall: ((VideoCallSession) -> Unit)? = null

    /** Fired when call state changes (answered, hung up, failed). */
    var onCallStateChanged: ((VideoCallSession) -> Unit)? = null

    /** Fired when a video frame arrives. */
    var onFrameReceived: ((UUID, ByteArray, Boolean) -> Unit)? = null

    /** Fired when the remote peer requests a keyframe. */
    var onKeyframeRequested: ((UUID) -> Unit)? = null

    /** Fired when the remote peer notifies a quality change. */
    var onQualityChanged: ((UUID, Int, Int, Int, Int) -> Unit)? = null

    // ─────────────────────────────────────────────────────────────────────
    // Outbound API
    // ─────────────────────────────────────────────────────────────────────

    /**
     * Initiate a video call to [toUhid]. Returns the new call-id.
     */
    suspend fun sendOffer(
        toUhid: String,
        codecs: List<String>,
        width: Int,
        height: Int,
        fps: Int,
        bitrateKbps: Int
    ): UUID {
        require(toUhid.isNotEmpty()) { "toUhid must not be empty" }
        val callId = UUID.randomUUID()
        val session = VideoCallSession(
            id = callId,
            callerUhid = sender.localUhid,
            calleeUhid = toUhid,
            state = VideoCallState.Outgoing,
            proposedCodecs = codecs,
            width = width,
            height = height,
            fps = fps,
            bitrateKbps = bitrateKbps
        )
        sessions[callId] = session

        val payload = encodeVideoSignaling(
            kind = "offer",
            callId = callId,
            fromUhid = sender.localUhid,
            toUhid = toUhid,
            proposedCodecs = codecs,
            width = width,
            height = height,
            fps = fps,
            bitrateKbps = bitrateKbps
        )
        sender.send(videoSignalingPacket(toUhid, payload), toUhid)
        return callId
    }

    /**
     * Accept an incoming call. Sends an "answer" signaling message.
     */
    suspend fun acceptCall(callId: UUID) {
        val session = sessions[callId] ?: return
        if (session.state != VideoCallState.Incoming) return
        session.state = VideoCallState.Connected
        onCallStateChanged?.invoke(session)

        val payload = encodeVideoSignaling(
            kind = "answer",
            callId = callId,
            fromUhid = sender.localUhid,
            toUhid = session.callerUhid,
            selectedCodec = session.selectedCodec,
            width = session.width,
            height = session.height,
            fps = session.fps,
            bitrateKbps = session.bitrateKbps
        )
        sender.send(videoSignalingPacket(session.callerUhid, payload), session.callerUhid)
    }

    /**
     * Hang up or cancel the call.
     */
    suspend fun hangUp(callId: UUID) {
        val session = sessions.remove(callId) ?: return
        val kind = if (session.state == VideoCallState.Outgoing) "cancel" else "hangup"
        val remote = if (sender.localUhid == session.callerUhid) session.calleeUhid else session.callerUhid
        session.state = VideoCallState.Ended
        onCallStateChanged?.invoke(session)

        val payload = encodeVideoSignaling(
            kind = kind,
            callId = callId,
            fromUhid = sender.localUhid,
            toUhid = remote
        )
        sender.send(videoSignalingPacket(remote, payload), remote)
    }

    /**
     * Send an encoded video frame for an active call.
     */
    suspend fun sendFrame(callId: UUID, encodedVideo: ByteArray, isKeyframe: Boolean) {
        val session = sessions[callId]
        if (session == null || session.state != VideoCallState.Connected) return
        val remote = if (sender.localUhid == session.callerUhid) session.calleeUhid else session.callerUhid

        val seq = frameSequence++.toUInt()
        val payload = encodeVideoFrame(callId, seq, System.currentTimeMillis(), isKeyframe, encodedVideo)
        val packet = MeshPacket(
            type = PacketType.VideoFrame,
            sourceUhid = sender.localUhid,
            destinationUhid = remote,
            ttl = AetherNetConstants.DEFAULT_TTL,
            priority = 64,
            payload = payload
        )
        sender.send(packet, remote)
    }

    /**
     * Request the remote peer send a keyframe (for decoder recovery).
     */
    suspend fun requestKeyframe(callId: UUID) {
        val session = sessions[callId] ?: return
        val remote = if (sender.localUhid == session.callerUhid) session.calleeUhid else session.callerUhid

        val payload = encodeVideoSignaling(
            kind = "keyframe_request",
            callId = callId,
            fromUhid = sender.localUhid,
            toUhid = remote
        )
        sender.send(videoSignalingPacket(remote, payload), remote)
    }

    /**
     * Notify the remote peer of a local quality/resolution change.
     */
    suspend fun notifyQualityChange(callId: UUID, width: Int, height: Int, fps: Int, bitrateKbps: Int) {
        val session = sessions[callId] ?: return
        val remote = if (sender.localUhid == session.callerUhid) session.calleeUhid else session.callerUhid
        session.width = width
        session.height = height
        session.fps = fps
        session.bitrateKbps = bitrateKbps

        val payload = encodeVideoSignaling(
            kind = "quality_change",
            callId = callId,
            fromUhid = sender.localUhid,
            toUhid = remote,
            width = width,
            height = height,
            fps = fps,
            bitrateKbps = bitrateKbps
        )
        sender.send(videoSignalingPacket(remote, payload), remote)
    }

    // ─────────────────────────────────────────────────────────────────────
    // Inbound packet dispatcher
    // ─────────────────────────────────────────────────────────────────────

    suspend fun handlePacket(packet: MeshPacket) {
        when (packet.type) {
            PacketType.VideoSignaling -> handleSignaling(packet)
            PacketType.VideoFrame -> handleFrame(packet)
            else -> {}
        }
    }

    // ─────────────────────────────────────────────────────────────────────
    // Private helpers
    // ─────────────────────────────────────────────────────────────────────

    private fun handleSignaling(packet: MeshPacket) {
        val json = String(packet.payload, Charsets.UTF_8)
        val kind = JsonReader.readString(json, "kind") ?: return
        val callIdStr = JsonReader.readString(json, "call_id") ?: return
        val callId = runCatching { UUID.fromString(callIdStr) }.getOrNull() ?: return
        val fromUhid = JsonReader.readString(json, "from_uhid") ?: return

        when (kind) {
            "offer" -> {
                val codecs = JsonReader.readStringArray(json, "proposed_codecs")
                val w = JsonReader.readInt(json, "width") ?: 0
                val h = JsonReader.readInt(json, "height") ?: 0
                val f = JsonReader.readInt(json, "fps") ?: 0
                val b = JsonReader.readInt(json, "bitrate_kbps") ?: 0
                val session = VideoCallSession(
                    id = callId,
                    callerUhid = fromUhid,
                    calleeUhid = sender.localUhid,
                    state = VideoCallState.Incoming,
                    proposedCodecs = codecs,
                    width = w,
                    height = h,
                    fps = f,
                    bitrateKbps = b
                )
                sessions[callId] = session
                onIncomingCall?.invoke(session)
            }
            "answer" -> {
                val session = sessions[callId] ?: return
                session.state = VideoCallState.Connected
                session.selectedCodec = JsonReader.readString(json, "selected_codec")
                onCallStateChanged?.invoke(session)
            }
            "hangup", "cancel", "timeout" -> {
                val session = sessions.remove(callId) ?: return
                session.state = VideoCallState.Ended
                onCallStateChanged?.invoke(session)
            }
            "keyframe_request" -> {
                onKeyframeRequested?.invoke(callId)
            }
            "quality_change" -> {
                val session = sessions[callId] ?: return
                val w = JsonReader.readInt(json, "width") ?: session.width
                val h = JsonReader.readInt(json, "height") ?: session.height
                val f = JsonReader.readInt(json, "fps") ?: session.fps
                val b = JsonReader.readInt(json, "bitrate_kbps") ?: session.bitrateKbps
                session.width = w; session.height = h; session.fps = f; session.bitrateKbps = b
                onQualityChanged?.invoke(callId, w, h, f, b)
            }
        }
    }

    private fun handleFrame(packet: MeshPacket) {
        val data = packet.payload
        // Minimum: 16 + 4 + 8 + 1 = 29 bytes
        if (data.size < 29) return

        val uuidBuf = ByteBuffer.wrap(data, 0, 16).order(ByteOrder.BIG_ENDIAN)
        val callId = UUID(uuidBuf.long, uuidBuf.long)

        val buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        buf.position(16)
        @Suppress("UNUSED_VARIABLE") val seq = buf.int.toUInt()
        @Suppress("UNUSED_VARIABLE") val ts = buf.long
        val isKeyframe = buf.get() != 0.toByte()
        val encoded = if (buf.hasRemaining()) {
            val arr = ByteArray(buf.remaining())
            buf.get(arr)
            arr
        } else ByteArray(0)

        onFrameReceived?.invoke(callId, encoded, isKeyframe)
    }

    private fun videoSignalingPacket(toUhid: String, payload: ByteArray) = MeshPacket(
        type = PacketType.VideoSignaling,
        sourceUhid = sender.localUhid,
        destinationUhid = toUhid,
        ttl = AetherNetConstants.DEFAULT_TTL,
        priority = 32,
        payload = payload
    )
}

// ─────────────────────────────────────────────────────────────────────────────
// Wire helpers
// ─────────────────────────────────────────────────────────────────────────────

internal fun encodeVideoFrame(
    callId: UUID,
    sequence: UInt,
    timestampMs: Long,
    isKeyframe: Boolean,
    encodedPayload: ByteArray
): ByteArray {
    val buf = ByteBuffer.allocate(16 + 4 + 8 + 1 + encodedPayload.size)
    buf.order(ByteOrder.BIG_ENDIAN)
    buf.putLong(callId.mostSignificantBits)
    buf.putLong(callId.leastSignificantBits)
    buf.order(ByteOrder.LITTLE_ENDIAN)
    buf.putInt(sequence.toInt())
    buf.putLong(timestampMs)
    buf.put(if (isKeyframe) 1.toByte() else 0.toByte())
    buf.put(encodedPayload)
    return buf.array()
}

private fun encodeVideoSignaling(
    kind: String,
    callId: UUID,
    fromUhid: String,
    toUhid: String,
    proposedCodecs: List<String>? = null,
    selectedCodec: String? = null,
    width: Int? = null,
    height: Int? = null,
    fps: Int? = null,
    bitrateKbps: Int? = null,
    reason: String? = null
): ByteArray {
    val sb = StringBuilder()
    sb.append('{')
    sb.append("\"kind\":\"").append(jsonEscape(kind)).append("\",")
    sb.append("\"call_id\":\"").append(callId).append("\",")
    sb.append("\"from_uhid\":\"").append(jsonEscape(fromUhid)).append("\",")
    sb.append("\"to_uhid\":\"").append(jsonEscape(toUhid)).append('"')
    if (proposedCodecs != null) {
        sb.append(",\"proposed_codecs\":[")
        proposedCodecs.forEachIndexed { i, c ->
            if (i > 0) sb.append(',')
            sb.append('"').append(jsonEscape(c)).append('"')
        }
        sb.append(']')
    }
    if (selectedCodec != null) sb.append(",\"selected_codec\":\"").append(jsonEscape(selectedCodec)).append('"')
    if (width != null) sb.append(",\"width\":").append(width)
    if (height != null) sb.append(",\"height\":").append(height)
    if (fps != null) sb.append(",\"fps\":").append(fps)
    if (bitrateKbps != null) sb.append(",\"bitrate_kbps\":").append(bitrateKbps)
    if (reason != null) sb.append(",\"reason\":\"").append(jsonEscape(reason)).append('"')
    sb.append('}')
    return sb.toString().toByteArray(Charsets.UTF_8)
}
